File: Hard/max_points_on_a_line/max_points_on_a_line.cpp
#include <vector>
#include <algorithm>
#include <iostream>
using namespace std;

class Solution{
    public:
        int maxPoints(vector<vector<int>> &points){
            int n = points.size();
            if(n <= 2){
                return n;
            }
            int maxPoints = 2;
            for(int i = 0; i < n; i++){
                for(int j = i+1; j < n; j++){
                    int count = 2;
                    int x1 = points[i][0], y1 = points[i][1], x2 = points[j][0], y2 = points[j][1];
                    for(int k = 0; k < n; k++){
                        if(k==i || k==j){
                            continue;
                        }
                        int x = points[k][0], y = points[k][1];
                        if((y2 - y1) * (x - x1) == (y - y1) * (x2 - x1)){
                            count++;
                        }
                    }
                    if(count > maxPoints){
                        maxPoints = count;
                    }
                }
            }
        return maxPoints;
    }
};

int main(){
    Solution sol;
    vector<vector<int>> pts1 = {{1,1},{2,2},{3,3}};
    vector<vector<int>> pts2 = {{1,1},{3,2},{5,3},{4,1},{2,3},{1,4}};

    cout << "For points {{1,1},{2,2},{3,3}}, max points on a line: " << sol.maxPoints(pts1) << endl;
    cout << "For points {{1,1},{3,2},{5,3},{4,1},{2,3},{1,4}}, max points on a line: " << sol.maxPoints(pts2) << endl;
    return 0;
}

File: Hard/max_points_on_a_line/max_points_on_a_line.py
from typing import List

def maxPoints(self, points: List[List[int]]) -> int:
    n = len(points)
    if n <= 2:
        return n
    max_points = 2

    for i in range(n):
        for j in range(i+1, n):
            count = 2 # points[i] and points[j] are already counted
            x1, y1 = points[i]
            x2, y2 = points[j]
            for k in range(n):
                if k == i or k == j:
                    continue
                x, y = points[k]
                # Check if points[i], points[j], and points[k] are collinear
                if (y2 - y1) * (x - x1) == (y - y1) * (x2 - x1):
                    count += 1
            if count > max_points:
                max_points = count
    return max_points

if __name__ == "__main__":
    print("Testing maxPoints function:")
    points = [[1,1],[2,2],[3,3]]
    print("For the input points =", points, ", the maximum number of points on a line is:", maxPoints(None, points))  # Expected output: 3

    points2 = [[1,1],[3,2],[5,3],[4,1],[2,3],[1,4]]
    print("For the input points =", points2, ", the maximum number of points on a line is:", maxPoints(None, points2))  # Expected output: 4